Domain fronting integration

durch opened the pull request:
Summary

This PR completes the migration from the NymApiClient to the nym_http_api_client::Client across the entire Nym monorepo, enabling domain fronting support and providing a consistent HTTP client interface throughout the codebase.

Changes

Core HTTP Client Unification

- Removed the Client re-export from nym_api module that was causing namespace conflicts
- Migrated all code from NymApiClient to the unified nym_http_api_client::Client
- Extended NymApiClientExt trait with missing methods to maintain backward compatibility:
    - get_current_rewarded_set() - for fetching epoch rewarded sets
    - get_all_basic_nodes_with_metadata() - for retrieving all network nodes
    - get_all_described_nodes() - for getting node descriptions
    - get_all_basic_active_mixing_assigned_nodes_with_metadata() - for active mixnodes
    - get_all_basic_entry_assigned_nodes_with_metadata() - for entry gateways
    - get_all_expanded_nodes() - for detailed node information
    - change_base_urls() - for API endpoint rotation
    - nodes_by_addresses() - for querying nodes by IP addresses

Domain Fronting Support

- Unified HTTP client now supports domain fronting configuration
- Client can rotate between multiple API endpoints for resilience (this needs extra review, I'm not sure its right)
- Proper URL type conversions implemented throughout

Crate Updates

- Added nym-http-api_client dependency to:
  - nym-validator-rewarder
  - nym-api
  - nym-network-monitor
  - nym-ecash-signer-check
  - testnet-manager
  - validator-status-check
  - nym-sdk
  - nym-client-core
- Added nym-mixnet-contract-common to nym-client-core for type conversions

Type System Improvements

- Fixed RewardedSetResponse → EpochRewardedSet → CachedEpochRewardedSet conversion chain
- Implemented proper error type conversions using From trait for NymAPIError
- Updated all error handling to use RequestError::new() constructor instead of direct field access

Component-Specific Changes

nym-node

  - Updated NodesQuerier to use unified client
  - Fixed API endpoint rotation logic in NymApisClient
  - Removed deprecated NymApiClient::from() usage
  - Updated broadcast functionality to work with new client

nym-api-provider

  - Refactored NymApiTopologyProvider to use unified client
  - Added support for multiple API URLs with automatic rotation on failure
  - Properly handles bincode encoding preference (though runtime toggle not yet supported)

validator-client

  - Consolidated client implementation to use unified HTTP client
  - Removed duplicate client instantiation code
  - Improved error handling and type safety

SDK and Examples

  - Updated custom_topology_provider example to use new client API
  - Fixed import paths and type conversions
  - Documentation updated to reflect new patterns

Error Handling

  - Consistent error construction using RequestError::new()
  - Proper error propagation with ? operator
  - Better error messages for debugging


Migration Notes

For code using the old NymApiClient:
```
// Before
let client = NymApiClient::new(url);
client.change_nym_api(new_url);

// After
let client = nym_http_api_client::Client::builder(url)?.build()?;
client.change_base_urls(vec![new_url]);

The NymApiClientExt trait must be imported to access Nym-specific methods:
use nym_validator_client::nym_api::NymApiClientExt;
```

Breaking Changes

- NymApiClient is now deprecated (though still available for backward compatibility)
- Client is no longer re-exported from nym_api module
- Some method signatures changed to use &mut self instead of &self (e.g., change_base_urls)

Future Work

- Complete removal of deprecated NymApiClient in next major version
- Runtime bincode toggle support in unified client
- Further optimization of paginated API calls

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/nymtech/nym/5974)
<!-- Reviewable:end -->

durch committed:
feat: unify HTTP client creation and enable domain fronting

Enhanced the base nym_http_api_client to reduce fragmentation and enable domain fronting:

- Added SerializationFormat enum for explicit JSON/bincode choice (no auto-detection)
- Added from_network() method to create clients from NymNetworkDetails with domain fronting
- Added with_bincode() builder method for explicit serialization configuration
- Set Accept header based on serialization preference
- Added deprecation paths for NymApiClient wrapper and nym_api::Client re-export
- Enabled domain fronting support via network defaults feature

This is part of a broader effort to consolidate HTTP client implementations across the codebase,
reducing ~500 lines of wrapper code and providing automatic domain fronting for censorship resistance.

durch committed:
feat: migrate NymApiClient usage to unified HTTP client

- Wire up domain fronting configuration in NymNetworkDetails
- Implement NymApiClientExt trait for base nym_http_api_client::Client
- Migrate direct NymApiClient usage in multiple components:
  - nym-network-monitor
  - verloc measurements
  - connection tester
  - coconut/ecash client
  - validator rewarder
- Add Copy derive to ApiUrlConst to enable iteration
- Update error handling and Display implementations

This enables automatic domain fronting for all Nym API calls via the configured CDN front hosts.

durch committed:
fix: resolve all compilation errors after NymApiClient migration

- Add missing nym-http-api-client dependencies to multiple crates
- Add NymApiClientExt trait imports where needed
- Fix type mismatches from NymApiClient to unified Client
- Add error conversions for NymAPIError in various error enums
- Implement missing trait methods (get_current_rewarded_set, get_all_basic_nodes_with_metadata, get_all_described_nodes)
- Fix type conversions for RewardedSetResponse in network monitor
- Update all API client instantiation to use new unified HTTP client

durch committed:
feat: complete migration to unified HTTP client and fix all compilation errors

- Added missing NymApiClientExt trait methods (get_all_expanded_nodes, change_base_urls)
- Fixed all compilation errors across the workspace
- Updated nym-node to use unified client instead of deprecated NymApiClient
- Fixed type conversions for RewardedSetResponse → EpochRewardedSet
- Added nym-http-api-client dependency where needed
- Updated all examples and documentation to use new client API

durch committed:
fix: provide all API URLs for automatic failover in endpoint rotation

Previously, when rotating API endpoints, only a single URL was provided to the
HTTP client, defeating the purpose of having multiple URLs for resilience.

Changes:
- NymApiTopologyProvider now provides all URLs in rotated order when switching endpoints
- NymApisClient similarly provides all URLs starting from the working endpoint
- Added clarifying comments for broadcast/exhaustive query methods where single URLs are intentionally used
- This enables the HTTP client's built-in failover mechanism while maintaining endpoint rotation behavior

The fix ensures that if the primary endpoint fails, the client can automatically
failover to alternative endpoints without manual intervention, improving overall
network resilience.

jstuczyn commented on common/http-api-client/src/lib.rs:
```diff
@@ -742,6 +824,13 @@ impl ApiClientCore for Client {
+        };
+        rb = rb.header(reqwest::header::ACCEPT, accept_header);
+
         if let Some(body) = json_body {
```
this would also have to be conditional in this instance. if we decided on `SerializationFormat::Bincode` then setting json body makes no sense

jstuczyn commented on common/http-api-client/src/lib.rs:
```diff
@@ -742,6 +824,13 @@ impl ApiClientCore for Client {
         let mut rb = RequestBuilder::from_parts(self.reqwest_client.clone(), req);
 
+        // Set Accept header based on serialization preference
+        let accept_header = match self.serialization {
```
would it perhaps be cleaner to add method on `SerializationFormat` to return `Option<&str>` (or whatever type is appropriate) for the header? especially if we decided to add more variants

jstuczyn commented on common/http-api-client/src/lib.rs:
```diff
@@ -191,6 +193,15 @@ pub type Params<'a, K, V> = &'a [(K, V)];
 
+/// Serialization format for API requests and responses
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum SerializationFormat {
```
if it's not too much of a hassle, for completion sake I'd also include yaml here (as it exists on `FormattedResponse`.

Perhaps also a good old `text` as sometimes you might want to send a string body without any further serialisation

jstuczyn commented on common/http-api-client/src/lib.rs:
```diff
@@ -742,6 +824,13 @@ impl ApiClientCore for Client {
+            SerializationFormat::Json => "application/json",
+            SerializationFormat::Bincode => "application/bincode",
+        };
+        rb = rb.header(reqwest::header::ACCEPT, accept_header);
```
would we also have to set `CONTENT_TYPE` (for anything other than `json`, as reqwest is already doing that)?

jstuczyn commented on common/http-api-client/src/lib.rs:
```diff
@@ -370,6 +381,7 @@ pub struct ClientBuilder {
     front: Option<fronted::Front>,
 
     retry_limit: usize,
+    serialization: SerializationFormat,
```
the more I think about it, the more I wonder: should it be property of a client or a request? because i could conceivably see a situation where you might want to connect to nym-api and requests some data as json and some as bincode

jstuczyn commented on common/http-api-client/src/lib.rs:
```diff
@@ -693,26 +764,37 @@ impl Client {
+
                 // this should never fail as we are transplanting the host from one url to another
-                r.url_mut().set_host(url.front_str()).unwrap();
+                r.url_mut().set_host(Some(front_host)).unwrap();
```
wouldn't this blow up if we have invalid hosts? especially since we do `.unwrap_or("")`

maybe add 
```toml
[lints]
workspace = true
```

to cargo.toml to catch all unwraps, expects, etc

jstuczyn commented on common/verloc/src/measurements/measurer.rs:
```diff
@@ -135,10 +135,17 @@ impl VerlocMeasurer {
-                self.config.user_agent.clone(),
-            );
+            let client =
+                match nym_http_api_client::Client::builder(api_endpoint.clone()).and_then(|b| {
```
given this instance is used by nym-node specifically, would it make sense to add a separate builder to the `Client` such that:
- it wouldn't be using hickory -> no point
- it wouldn't be using domain fronting
?

durch replied:
Maybe, but then its another single purpose builder, I think we should dogfood this client as much as possible, will find nastiness sooner ;)

jstuczyn commented on common/client-core/src/client/base_client/mod.rs:
```diff
@@ -749,21 +749,42 @@ where
+                source: nym_validator_client::nym_api::error::NymAPIError::GenericRequestFailure(
+                    e.to_string(),
+                ),
+            })
```
```suggestion
        builder
            .with_bincode()
            .build::<nym_validator_client::models::RequestError>()
            .map_err(|e| ClientCoreError::NymApiQueryFailure {
                source: nym_validator_client::nym_api::error::NymAPIError::GenericRequestFailure(
                    e.to_string(),
                ),
            })
```

jstuczyn commented on common/http-api-client/src/lib.rs:
```diff
@@ -500,6 +557,17 @@ impl ClientBuilder {
+
+    /// Configure the client to use bincode serialization
+    pub fn with_bincode(self) -> Self {
+        self.with_serialization(SerializationFormat::Bincode)
```
what if the target host doesn't support bincode? though I guess it's the caller's fault for being stupid for trying to use it in the first place

durch replied:
Yeah, thats a risk, but some responsibility has to be with the caller

jstuczyn commented on common/client-core/src/init/helpers.rs:
```diff
@@ -83,6 +84,48 @@ struct GatewayWithLatency<'a, G: ConnectableGateway> {
 }
 
+// Helper to collect all pages of entry nodes - replicates NymApiClient's convenience method
+async fn get_all_basic_entry_nodes_with_metadata(
```
given you had to extract it, maybe we should also do a tiny cleanup of `NymApiClient` and move all of those methods to some `Ext` trait on the `nym_http_api_client::Client`?

durch replied:
I'd leave that for cleanup number two, there are a bunch of similar tweaks we should do

jstuczyn commented on common/client-libs/validator-client/src/nym_api/mod.rs:
```diff
@@ -241,6 +243,162 @@ pub trait NymApiClientExt: ApiClient {
+        self.get_rewarded_set().await
+    }
+
+    async fn get_all_basic_nodes_with_metadata(
```
would the `collect_paged_skimmed_v2` macro help here?

durch replied:
Oh nice, I've got a bunch of these, I'll ticket them up and clean it!

jstuczyn reviewed (commented):
looks good! only some minor comments

durch committed:
Update common/client-core/src/client/base_client/mod.rs

Co-authored-by: Jędrzej Stuczyński <jedrzej.stuczynski@gmail.com>

durch committed:
Remove error generics, address PR comments

jstuczyn commented on common/http-api-client/Cargo.toml:
```diff
@@ -23,6 +24,8 @@ url = { workspace = true }
 serde = { workspace = true }
 serde_json = { workspace = true }
+serde_yaml = "0.9.34-deprecated"
+serde_plain = "1.0.2"
```
let's not pollute our crates with standalone dependency versioning. could you set those verisons in the workspace Cargo.toml? 🙏🏻

durch committed:
Explicit warning on missing fronting configuration

durch committed:
Merge branch 'develop' into domain-fronting-integration

durch committed:
Assorted CI fixes

durch committed:
Merge branch 'domain-fronting-integration' of https://github.com/nymtech/nym into domain-fronting-integration

durch committed:
Merge branch 'develop' into domain-fronting-integration

neacsu committed:
Use default value for the ports until api is deployed

durch committed:
Registry proc-macro

durch committed:
Rename macro

durch committed:
Syn workspace version

durch committed:
Where do we need to put inventory

durch committed:
Ergonomics and call sites, incept the builder

durch committed:
fix: Address critical issues in client configuration registry implementation

- Fixed HeaderMapInit parsing bug that would cause compilation errors
- Added comprehensive documentation with usage examples and DSL reference
- Improved error handling with better error messages for invalid headers
- Added test coverage for both macro and registry functionality
- Added debug inspection capabilities for registered configurations
- Fixed module name conflicts in tests by using separate modules

All tests now passing:
- 7 macro tests validating DSL parsing and code generation
- 4 registry tests verifying configuration collection and application

jstuczyn committed:
Feature/improved http error (#6025)

* use display impl for urls

* feat: attempt to add more details to reqwest errors

* temporarily restored GenericRequestFailure variant

* another restoration

* cleanup

durch committed:
Some debug tooling, and default timeout fix

durch committed:
Fix user-agent override

durch committed:
Merge branch 'develop' into domain-fronting-integration

durch committed:
Merge branch 'drazen/client-configuration-registry' into domain-fronting-integration

durch committed:
Fix various wasm things
